app/models/talk/searchable.rb

```diff
@@ -4,62 +4,37 @@ module Talk::Searchable
   DATE_WEIGHT = 0.000000001
 
   included do
-    scope :ft_search, ->(query) do
-      query = query&.gsub(/[^[:word:]]/, " ") || "" # remove non-word characters
-      query = query.split.map { |word| "#{word}*" }.join(" ") # wildcard search
-      joins("join talks_search_index idx on talks.id = idx.rowid")
-        .where("talks_search_index match ?", query)
-    end
+    has_one :index, foreign_key: :rowid, dependent: :destroy
+
+    scope :ft_search, ->(query) { select("talks.*").joins(:index).merge(Talk::Index.search(query)) }
 
     scope :with_snippets, ->(**options) do
-      select("talks.*")
-        .select_snippet("title", 0, **options)
-        .select_snippet("summary", 1, **options)
-        .select_snippet("speaker_names", 2, **options)
+      select("talks.*").merge(Talk::Index.snippets(**options))
     end
 
     scope :ranked, -> do
       select("talks.*,
           bm25(talks_search_index, 10.0, 1.0, 5.0) +
           (strftime('%s', 'now') - strftime('%s', talks.date)) * #{DATE_WEIGHT} AS combined_score")
-        .order(Arel.sql("combined_score ASC"))
+        .order(combined_score: :asc)
     end
 
-    after_create_commit :create_in_index
-    after_update_commit :update_in_index
-    after_destroy_commit :remove_from_index
+    after_save_commit :reindex
   end
 
   class_methods do
-    def rebuild_search_index
-      connection.execute("DELETE FROM talks_search_index")
-      Talk.find_each(&:create_in_index)
-    end
-
-    def select_snippet(column, offset, tag: "mark", omission: "…", limit: 32)
-      select("snippet(talks_search_index, #{offset}, '<#{tag}>', '</#{tag}>', '#{omission}', #{limit}) AS #{column}_snippet")
+    def reindex_all
+      Talk::Index.delete_all
```

Comment:
@adrienpoly reading this again I'm confused as to why we need to delete the existing ones beforehand? Do you remember?

Reply:
no I don't remember. Maybe I did that initially to just make sure we get something clean knowing that our data is verylimieted so doing a full rebuild is not really costly

Reply:
@adrienpoly ah got it. Yeah, I think with this change we could let Active Record decide and spare some needless queries in case there's no update needed.

Reply:
```suggestion
```